apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/Keys.cs

```diff
@@ -275,49 +255,56 @@ private V1.KeyRange ToProtobuf()
     /// </summary>
     public sealed class Key
     {
-        /// <summary>
-        /// The values of this key.
-        /// </summary>
-        internal ListValue KeyParts { get; }
+        internal SpannerParameterCollection Parameters { get; }
 
         /// <summary>
         /// Creates a key from the values in the parameter collection.
         /// </summary>
         /// <param name="keyParts">The values that define the key</param>
         internal Key(SpannerParameterCollection keyParts)
         {
-            KeyParts = ToListValue(keyParts);
+            Parameters = keyParts;
         }
 
         /// <summary>
         /// Creates a key consisting of the given key parts.
         /// </summary>
         /// <param name="keyParts">The values that define the key</param>
-        public Key(params object[] keyParts)
+        public Key(params object[] keyParts) : this(new SpannerParameterCollection(keyParts.Select(j => new SpannerParameter { Value = j })))
         {
-            var protobufValues = keyParts.Select(
-                v => SpannerDbType.FromClrType(v?.GetType())?.ToProtobufValue(v) ?? Value.ForNull());
-            KeyParts = new ListValue { Values = { protobufValues } };
         }
 
-        /// <inheritdoc/>
-        public override bool Equals(object other)
-            => other is Key key && KeyParts.Equals(key.KeyParts);
+        /// <summary>
+        /// Gets the values of this Key.
+        /// </summary>
+        /// <param name="options">The options for conversion between protobuf and CLR types.</param>
+        /// <returns>The values of this Key.</returns>
+        internal ListValue GetKeyParts(SpannerConversionOptions options) => ToProtobuf(options);
 
-        /// <inheritdoc/>
-        public override int GetHashCode() => KeyParts.GetHashCode();
+        /// <summary>
+        /// Returns the string representation of this key using the default type conversion mappings.
+        /// The string representation is only for diagnostic purposes and the implementation may change over time.
+        /// </summary>
+        /// <returns>A string that represents the current key.</returns>
+        public override string ToString() => ToProtobuf(SpannerConversionOptions.Default).ToString();
```

Comment:
Hmmm, using Default here won't match the actual values sent to Spanner and I wonder if that would hinder diagnostics which is what this method will most likely be used for.

It was actually @jskeet who requested the ToString initially (https://github.com/googleapis/google-cloud-dotnet/pull/7157#discussion_r729612411) so we know it's not a requierement. I say we remove it. What do you think @jskeet .

Reply:
Yep, I still think removing is the better option right now, and adding the overload with the parameter to all three classes.

Reply:
Thanks Jon and Amanda . I will remove the override of ToString() and add ToString(SpannerConnectionStringBuilder),  update the tests and raise the commit by tomorrow morning.